ibtk/include/ibtk/PETScMatUtilities.h

```diff
@@ -159,18 +130,40 @@ class PETScMatUtilities
                                               int dof_index_idx,
                                               SAMRAI::tbox::Pointer<SAMRAI::hier::PatchLevel<NDIM> > patch_level);
 
-    static inline void ib_4_interp_fcn(const double r, double* const w)
+    static void ib_4_interp_fcn(const double r, double* const w)
     {
-        const double q = sqrt(1.0 + 4.0 * r * (1.0 - r));
-        w[0] = 0.125 * (3.0 - 2.0 * r - q);
-        w[1] = 0.125 * (3.0 - 2.0 * r + q);
-        w[2] = 0.125 * (1.0 + 2.0 * r + q);
-        w[3] = 0.125 * (1.0 + 2.0 * r - q);
+        const double q = sqrt(-7.0 + 12.0 * r - 4.0 * r * r);
+        w[0] = 0.125 * (5.0 - 2.0 * r - q);
+        w[1] = 0.125 * (5.0 - 2.0 * r + q);
+        w[2] = 0.125 * (-1.0 + 2.0 * r + q);
+        w[3] = 0.125 * (-1.0 + 2.0 * r - q);
         return;
     } // ib_4_interp_fcn
 
     static const int ib_4_interp_stencil = 4;
 
+    /*!
+     * \brief Construct a parallel PETSc Mat object corresponding to data
+     * prolongation from a coarser level to a finer level.
+     */
+    static void
+    constructPatchLevelProlongationOp(Mat& mat,
+                                      int dof_index_idx,
+                                      const std::vector<int>& num_fine_dofs_per_proc,
+                                      const std::vector<int>& num_coarse_dofs_per_proc,
+                                      SAMRAI::tbox::Pointer<SAMRAI::hier::PatchLevel<NDIM> > fine_patch_level,
+                                      SAMRAI::tbox::Pointer<SAMRAI::hier::PatchLevel<NDIM> > coarse_patch_level,
+                                      const AO& coarse_level_ao,
+                                      const int coarse_ao_offset = 0);
+
+    /*!
+     * \brief Construct a diagonal matrix that scales the rows of adjoint of
```

Comment:
On Tue, Apr 21, 2015 at 10:22 AM, Boyce Griffith <notifications@github.com>
wrote:

> In ibtk/include/ibtk/PETScMatUtilities.h
> <https://github.com/IBAMR/IBAMR/pull/6#discussion_r28788672>:
>
> > +    /*!
> > +     * \brief Construct a parallel PETSc Mat object corresponding to data
> > +     * prolongation from a coarser level to a finer level.
> > +     */
> > +    static void
> > +    constructPatchLevelProlongationOp(Mat& mat,
> > +                                      int dof_index_idx,
> > +                                      const std::vector<int>& num_fine_dofs_per_proc,
> > +                                      const std::vector<int>& num_coarse_dofs_per_proc,
> > +                                      SAMRAI::tbox::Pointer<SAMRAI::hier::PatchLevel<NDIM> > fine_patch_level,
> > +                                      SAMRAI::tbox::Pointer<SAMRAI::hier::PatchLevel<NDIM> > coarse_patch_level,
> > +                                      const AO& coarse_level_ao,
> > +                                      const int coarse_ao_offset = 0);
> > +
> > +    /*!
> > +     * \brief Construct a diagonal matrix that scales the rows of adjoint of
>
> Adjoint or transpose? Transpose, right?
>
For real matrices, they are the same.

  Matt

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/IBAMR/IBAMR/pull/6/files#r28788672>.
>



-- 
What most experimenters take for granted before they begin their
experiments is infinitely more interesting than any results to which their
experiments lead.
-- Norbert Wiener